substratevm/src/com.oracle.svm.agent/src/com/oracle/svm/agent/BreakpointInterceptor.java

```diff
@@ -913,6 +922,123 @@ private static String asInternalSignature(Object paramTypesArray) {
         return null;
     }
 
+    private static boolean objectStreamClassConstructor(JNIEnvironment jni, Breakpoint bp) {
+        JNIObjectHandle callerClass = getDirectCallerClass();
+        JNIObjectHandle self = getObjectArgument(0);
+        JNIObjectHandle serializeTargetClass = getObjectArgument(1);
+        String serializeTargetClassName = getClassNameOrNull(jni, serializeTargetClass);
+        long checksum = 0;
+        List<SerializationInfo> traceCandidates = new ArrayList<>();
+        JNIObjectHandle javaIOObjectStreamCaches = agent.handles().findClass(jni, "java/io/ObjectStreamClass$Caches");
+        JNIFieldId localDescsFId = agent.handles().getFieldId(jni, javaIOObjectStreamCaches, "localDescs", "Ljava/util/concurrent/ConcurrentMap;", true);
+        JNIObjectHandle concurrentHashMap = agent.handles().findClass(jni, "Ljava/util/concurrent/ConcurrentHashMap;");
+
+        /** Get ObjectStreamClass.Caches.localDescs' size before init */
+        JNIObjectHandle localDescs = jniFunctions().getGetStaticObjectField().invoke(jni, javaIOObjectStreamCaches, localDescsFId);
+        JNIMethodId sizeMId = agent.handles().getMethodId(jni, concurrentHashMap, "size", "()I", false);
+        int oldCacheSize = jniFunctions().getCallIntMethodA().invoke(jni, localDescs, sizeMId, nullPointer());
+
+        JNIObjectHandle objectStreamClassInstance = newObjectL(jni, bp.clazz, bp.method, serializeTargetClass);
+        Object result = nullHandle().notEqual(objectStreamClassInstance);
+        if (clearException(jni)) {
+            result = false;
+        }
+        if (result.equals(true)) {
+            checksum = getChecksumFromOSC(jni, bp.clazz, objectStreamClassInstance, serializeTargetClass, serializeTargetClassName);
+        }
+        traceCandidates.add(new SerializationInfo(serializeTargetClassName, checksum));
+
+        /**
+         * Check if ObjectStreamClass.Caches.localDescs has been changed If there are more elements
+         * after the execution of ObjectStreamClass.<init>, it suggests the super class(es) of
+         * current serialization target class get(s) serialized recursively during the process of
+         * initialization. We need to find out the corresponding classes and checksums and add them
+         * to the serialization configurations.
+         */
```

Comment:
If I understand correctly, the actions in this method can cause other `ObjectStreamClass` instances to be initialized and we miss that because the breakpoint is not triggered recursively. In that case, we should change the breakpoint to be in `ObjectStreamClass.lookup(Class<?>, boolean)` instead, which is called each time when a class is used in serialization, so we should never miss a class. For efficiency, we can just check in a map whether we have already handled this class, but we should still emit an event (with cached data) and do the access check with `serializationAccessVerifier`. This should be simpler and more reliable than trying to infer those classes from the descriptor cache.

Reply:
Method `ObjectStreamClass.lookup(Class<?>, boolean)` is still recursively called during serialization. Given the followsing class hierarchy for example:
![image](https://user-images.githubusercontent.com/6834162/96686848-a72fd100-13b1-11eb-864c-b35e1dbd9940.png)
It has the following call chain when serialize the `Child` object. The `lookup` method is recursively invoked for each class along the class hierarchy. So it is no need to change to set break point on `lookup` method.
![image](https://user-images.githubusercontent.com/6834162/96685966-700cf000-13b0-11eb-9e55-2d0e974c012f.png)

Reply:
Fixed. `dataLayout` is null until method `ObjectStreamClass.getClassDataLayout` is called. I call method `ObjectStreamClass.getClassDataLayout0()` which returns an array of `ClassDataSlot` in which all necessary data can be found but doesn't fill `dataLayout`.